nexus/db-queries/src/db/datastore/sled.rs

```diff
@@ -287,6 +288,308 @@ fn pick_sled_reservation_target(
+            // request. These will be filtered later.
+            let candidate_datasets: HashSet<CandidateDataset> = zpools_for_sled
+                .iter()
+                .filter(|zpool_get_result| {
+                    let ZpoolGetForSledReservationResult {
+                        pool,
+                        last_inv_total_size,
+                        rendezvous_local_storage_dataset_id: _,
+                        crucible_dataset_usage,
+                        local_storage_usage,
+                    } = zpool_get_result;
+
+                    // The total request size for the local storage dataset
+                    // allocation is the disk size plus the required
+                    // overhead.
+                    let request_size: i64 = request.size().to_bytes() as i64
+                        + request.required_dataset_overhead().to_bytes() as i64;
+
+                    let new_size_used: i64 = crucible_dataset_usage
+                        + local_storage_usage
+                        + request_size;
+
+                    let control_plane_storage_buffer: i64 =
+                        pool.control_plane_storage_buffer().into();
+                    let adjusted_total: i64 =
+                        last_inv_total_size - control_plane_storage_buffer;
+
+                    // Any zpool that has space for this local storage
+                    // dataset allocation is considered a candidate.
+                    new_size_used < adjusted_total
+                })
+                .map(|zpool_get_result| CandidateDataset {
+                    rendezvous_local_storage_dataset_id: zpool_get_result
+                        .rendezvous_local_storage_dataset_id,
+                    pool_id: zpool_get_result.pool.id(),
+                    sled_id: zpool_get_result.pool.sled_id(),
+                })
+                .collect();
+
+            if candidate_datasets.is_empty() {
+                // if there's no local storage datasets on this sled for this
+                // request's size, then try another sled.
+                info!(
+                    &log,
+                    "sled {sled_target} does not have any candidate datasets \
+                    with available space to satisfy local storage allocation";
+                    "request" => ?request,
+                );
+                return Err(());
+            }
+
+            allocations_to_perform.push(PossibleAllocationsForRequest {
+                request,
+                candidate_datasets,
+            });
+        }
+
+        // From the list of allocations to perform, and all the candidate local
+        // storage datasets that could fit those allocations, find a list of all
+        // valid request -> zpool mappings.
+        //
+        // Start from no allocations made yet, a list of allocations to perform
+        // (stored in `requests`), and all of the available zpools on the sled.
+        // Use a BinaryHeap to get a sorted list.
```

Comment:
maybe worth saying "sorted by the number of possible allocations", since otherwise, you have to go and track down the `Ord` impl to figure that out?